File: test/elixir/hologram/ex_js_consistency/elixir/string/tokenizer_test.exs
defmodule Hologram.ExJsConsistency.Elixir.String.TokenizerTest do
  @moduledoc """
  IMPORTANT!
  Each Elixir consistency test has a related JavaScript test in test/javascript/elixir/string/tokenizer_test.mjs
  Always update both together.
  """

  use Hologram.Test.BasicCase, async: true

  @moduletag :consistency

  describe "tokenize/1" do
    test "lowercase identifier" do
      assert String.Tokenizer.tokenize(~c"foo") == {:identifier, ~c"foo", [], 3, true, []}
    end

    test "identifier with underscore and digits" do
      assert String.Tokenizer.tokenize(~c"foo_bar1") ==
               {:identifier, ~c"foo_bar1", [], 8, true, []}
    end

    test "identifier starting with underscore" do
      assert String.Tokenizer.tokenize(~c"_foo") == {:identifier, ~c"_foo", [], 4, true, []}
    end

    test "identifier closed by punctuation" do
      assert String.Tokenizer.tokenize(~c"foo?") ==
               {:identifier, ~c"foo?", [], 4, true, [:punctuation]}
    end

    test "punctuation closes the identifier before the rest" do
      assert String.Tokenizer.tokenize(~c"foo??") ==
               {:identifier, ~c"foo?", ~c"?", 4, true, [:punctuation]}
    end

    test "identifier carrying the at sign" do
      assert String.Tokenizer.tokenize(~c"foo@bar") ==
               {:identifier, ~c"foo@bar", [], 7, true, [:at]}
    end

    test "alias" do
      assert String.Tokenizer.tokenize(~c"FooBar") == {:alias, ~c"FooBar", [], 6, true, []}
    end

    test "alias stops at the dot" do
      assert String.Tokenizer.tokenize(~c"Elixir.Foo") ==
               {:alias, ~c"Elixir", ~c".Foo", 6, true, []}
    end

    test "identifier stops at an ASCII non-identifier character" do
      assert String.Tokenizer.tokenize(~c"foo bar") ==
               {:identifier, ~c"foo", ~c" bar", 3, true, []}
    end

    test "Unicode identifier" do
      assert String.Tokenizer.tokenize(~c"héllo") == {:identifier, ~c"héllo", [], 5, false, []}
    end

    test "Han identifier" do
      assert String.Tokenizer.tokenize(~c"日本語") == {:identifier, ~c"日本語", [], 3, false, []}
    end

    test "uppercase non-ASCII starts an atom" do
      assert String.Tokenizer.tokenize(~c"Ω") == {:atom, ~c"Ω", [], 1, false, []}
    end

    test "NFC-normalizes and flags an unstable identifier" do
      # e followed by combining acute normalizes to é
      assert String.Tokenizer.tokenize([101, 769]) ==
               {:identifier, ~c"é", [], 2, false, [:nfkc]}
    end

    test "micro sign normalizes to Greek mu" do
      assert String.Tokenizer.tokenize([181]) == {:identifier, ~c"μ", [], 1, false, [:nfkc]}
    end

    test "Greek mu combines with any script" do
      assert String.Tokenizer.tokenize(~c"aμ") == {:identifier, ~c"aμ", [], 2, false, []}
    end

    test "underscore separates chunks checked for scripts on their own" do
      assert String.Tokenizer.tokenize(~c"fox_狐") == {:identifier, ~c"fox_狐", [], 5, false, []}
    end

    test "empty input" do
      assert String.Tokenizer.tokenize(~c"") == {:error, :empty}
    end

    test "digit can't start an identifier" do
      assert String.Tokenizer.tokenize(~c"1abc") == {:error, :empty}
    end

    test "at sign can't start an identifier" do
      assert String.Tokenizer.tokenize(~c"@foo") == {:error, :empty}
    end

    test "restricted codepoint can't start an identifier" do
      # fullwidth f is excluded by UTS 39
      assert String.Tokenizer.tokenize(~c"ｆｏｏ") == {:error, :empty}
    end

    test "restricted codepoint inside an identifier is unexpected" do
      # superscript two is excluded by UTS 39
      assert String.Tokenizer.tokenize(~c"x²") == {:error, {:unexpected_token, ~c"x²"}}
    end

    test "mixed scripts are rejected" do
      assert {:error, {:mixed_script, ~c"a日", {prefix, suffix}}} =
               String.Tokenizer.tokenize(~c"a日")

      # The prefix matches the client. The client's suffix explanation is simplified - it lists
      # the characters without their script names.
      assert prefix == ~c"invalid mixed-script identifier found: "
      assert is_list(suffix)
    end

    test "raises FunctionClauseError if the argument is not a list" do
      expected_msg =
        build_function_clause_error_msg("String.Tokenizer.tokenize/1", [:abc], [
          "def tokenize(-[head | tail]-)",
          "def tokenize(-[]-)"
        ])

      assert_error FunctionClauseError, expected_msg, fn ->
        :abc
        |> wrap_term()
        |> String.Tokenizer.tokenize()
      end
    end
  end
end

File: test/javascript/elixir/string/tokenizer_test.mjs
"use strict";

import {
  assert,
  assertBoxedError,
  defineRuntimeGlobals,
} from "../../support/helpers.mjs";

import Elixir_String_Tokenizer from "../../../../assets/js/elixir/string/tokenizer.mjs";
import Interpreter from "../../../../assets/js/interpreter.mjs";
import Type from "../../../../assets/js/type.mjs";

defineRuntimeGlobals();

// Mirrors the clause heads the compiler emits for the ported function, which
// the runtime script registers when the bundle loads.
Interpreter.defineFunctionClauseHeads(
  "String.Tokenizer",
  "tokenize",
  1,
  "public",
  [
    {
      params: (_context) => [
        Type.consPattern(
          Type.variablePattern("head_0"),
          Type.variablePattern("tail_1"),
        ),
      ],
      guards: [],
      blame: {params: ["[head | tail]"], guards: []},
    },
    {
      params: (_context) => [Type.list()],
      guards: [],
      blame: {params: ["[]"], guards: []},
    },
  ],
);

const tokenize = (codePoints) =>
  Elixir_String_Tokenizer["tokenize/1"](
    Type.list(codePoints.map((codePoint) => Type.integer(codePoint))),
  );

const codePoints = (text) => Array.from(text).map((c) => c.codePointAt(0));

const charlist = (text) => Type.list(codePoints(text).map(Type.integer));

const success = (kind, acc, rest, length, asciiLetters, special) =>
  Type.tuple([
    Type.atom(kind),
    charlist(acc),
    charlist(rest),
    Type.integer(length),
    Type.boolean(asciiLetters),
    Type.list(special.map(Type.atom)),
  ]);

// IMPORTANT!
// Each JavaScript test has a related Elixir consistency test in test/elixir/hologram/ex_js_consistency/elixir/string/tokenizer_test.exs
// Always update both together.

describe("Elixir_String_Tokenizer", () => {
  describe("tokenize/1", () => {
    it("lowercase identifier", () => {
      assert.deepStrictEqual(
        tokenize(codePoints("foo")),
        success("identifier", "foo", "", 3, true, []),
      );
    });

    it("identifier with underscore and digits", () => {
      assert.deepStrictEqual(
        tokenize(codePoints("foo_bar1")),
        success("identifier", "foo_bar1", "", 8, true, []),
      );
    });

    it("identifier starting with underscore", () => {
      assert.deepStrictEqual(
        tokenize(codePoints("_foo")),
        success("identifier", "_foo", "", 4, true, []),
      );
    });

    it("identifier closed by punctuation", () => {
      assert.deepStrictEqual(
        tokenize(codePoints("foo?")),
        success("identifier", "foo?", "", 4, true, ["punctuation"]),
      );
    });

    it("punctuation closes the identifier before the rest", () => {
      assert.deepStrictEqual(
        tokenize(codePoints("foo??")),
        success("identifier", "foo?", "?", 4, true, ["punctuation"]),
      );
    });

    it("identifier carrying the at sign", () => {
      assert.deepStrictEqual(
        tokenize(codePoints("foo@bar")),
        success("identifier", "foo@bar", "", 7, true, ["at"]),
      );
    });

    it("alias", () => {
      assert.deepStrictEqual(
        tokenize(codePoints("FooBar")),
        success("alias", "FooBar", "", 6, true, []),
      );
    });

    it("alias stops at the dot", () => {
      assert.deepStrictEqual(
        tokenize(codePoints("Elixir.Foo")),
        success("alias", "Elixir", ".Foo", 6, true, []),
      );
    });

    it("identifier stops at an ASCII non-identifier character", () => {
      assert.deepStrictEqual(
        tokenize(codePoints("foo bar")),
        success("identifier", "foo", " bar", 3, true, []),
      );
    });

    it("Unicode identifier", () => {
      assert.deepStrictEqual(
        tokenize(codePoints("héllo")),
        success("identifier", "héllo", "", 5, false, []),
      );
    });

    it("Han identifier", () => {
      assert.deepStrictEqual(
        tokenize(codePoints("日本語")),
        success("identifier", "日本語", "", 3, false, []),
      );
    });

    it("uppercase non-ASCII starts an atom", () => {
      assert.deepStrictEqual(
        tokenize(codePoints("Ω")),
        success("atom", "Ω", "", 1, false, []),
      );
    });

    it("NFC-normalizes and flags an unstable identifier", () => {
      // e followed by combining acute normalizes to é
      assert.deepStrictEqual(
        tokenize([101, 769]),
        success("identifier", "é", "", 2, false, ["nfkc"]),
      );
    });

    it("Greek mu combines with any script", () => {
      assert.deepStrictEqual(
        tokenize(codePoints("aμ")),
        success("identifier", "aμ", "", 2, false, []),
      );
    });

    it("underscore separates chunks checked for scripts on their own", () => {
      assert.deepStrictEqual(
        tokenize(codePoints("fox_狐")),
        success("identifier", "fox_狐", "", 5, false, []),
      );
    });

    it("empty input", () => {
      assert.deepStrictEqual(
        tokenize([]),
        Type.tuple([Type.atom("error"), Type.atom("empty")]),
      );
    });

    it("digit can't start an identifier", () => {
      assert.deepStrictEqual(
        tokenize(codePoints("1abc")),
        Type.tuple([Type.atom("error"), Type.atom("empty")]),
      );
    });

    it("at sign can't start an identifier", () => {
      assert.deepStrictEqual(
        tokenize(codePoints("@foo")),
        Type.tuple([Type.atom("error"), Type.atom("empty")]),
      );
    });

    it("restricted codepoint can't start an identifier", () => {
      // fullwidth f is excluded by UTS 39
      assert.deepStrictEqual(
        tokenize(codePoints("ｆｏｏ")),
        Type.tuple([Type.atom("error"), Type.atom("empty")]),
      );
    });

    it("restricted codepoint inside an identifier is unexpected", () => {
      // superscript two is excluded by UTS 39
      assert.deepStrictEqual(
        tokenize(codePoints("x²")),
        Type.tuple([
          Type.atom("error"),
          Type.tuple([Type.atom("unexpected_token"), charlist("x²")]),
        ]),
      );
    });

    it("mixed scripts are rejected", () => {
      const result = tokenize(codePoints("a日"));
      const [tag, reason] = result.data;

      assert.deepStrictEqual(tag, Type.atom("error"));

      const [reasonTag, acc, message] = reason.data;

      assert.deepStrictEqual(reasonTag, Type.atom("mixed_script"));
      assert.deepStrictEqual(acc, charlist("a日"));

      // The prefix matches the server. The suffix explanation is simplified - it lists the
      // characters without their script names (see the module comment).
      assert.deepStrictEqual(
        message.data[0],
        charlist("invalid mixed-script identifier found: "),
      );

      assert.isTrue(Type.isList(message.data[1]));
    });

    it("raises FunctionClauseError if the argument is not a list", () => {
      assertBoxedError(
        () => Elixir_String_Tokenizer["tokenize/1"](Type.atom("abc")),
        "FunctionClauseError",
        "no function clause matching in String.Tokenizer.tokenize/1\n\nThe following arguments were given to String.Tokenizer.tokenize/1:\n\n    # 1\n    :abc\n\nAttempted function clauses (showing 2 out of 2):\n\n    def tokenize(-[head | tail]-)\n    def tokenize(-[]-)\n",
      );
    });
  });
});
